src/paginator-spec.js

```diff
@@ -0,0 +1,486 @@
+      });
+    });
+
+    describe('constructor', function() {
+      const givenPageNumber = 42;
+
+      beforeEach(function() {
+        const queryParams = {
+          page: givenPageNumber
+        };
+        paginator = new Paginator(request.spy, null, queryParams);
+      });
+
+      it('sets the page to the given page query parameter', function() {
+        expect(paginator.page).to.equal(givenPageNumber);
+      });
+    });
+
+    describe('constructor', function() {
+      const message = Paginator.errors.invalidStyle;
+      const style = 'invalid-value';
+
+      it('throws an error when an invalid pagination style is set',
+        function() {
+          const paginatorConstructor = function() {
+            new Paginator(request.spy, null, {}, { style });
+          };
+
+          return expect(paginatorConstructor).to.throw(Error, message);
+        });
+    });
+
+    describe('constructor', function() {
+      describe('(LimitOffsetPagination)', function() {
+        const style = Paginator.styles.limitOffset;
+
+        it('throws an error when no limit is set and offset isn\'t zero',
+          function() {
+            const message = Paginator.errors.calculatePageNoLimit;
+            const queryParams = { offset: 5 };
+            const paginatorConstructor = function() {
+              new Paginator(request.spy, null, queryParams, { style });
+            };
+
+            return expect(paginatorConstructor).to.throw(Error, message);
+          });
+
+        it('doesn\'t throw an error when no limit is set and offset is zero',
+          function() {
+            const message = Paginator.errors.calculatePageNoLimit;
+            const queryParams = { offset: 0 };
+            const paginatorConstructor = function() {
+              new Paginator(request.spy, null, queryParams, { style });
+            };
+
+            return expect(paginatorConstructor).to.not.throw(Error, message);
+          });
+      });
+    });
+
+    describe('page', function() {
+      beforeEach(function() {
+        paginator = new Paginator(request.spy);
+      });
+
+      it('defaults to one', function() {
+        expect(paginator.page).to.equal(1);
+      });
+    });
+
+    describe('next', function() {
+      beforeEach(function() {
+        paginator = new Paginator(request.spy);
+      });
+
+      it('returns a promise that successfully resolves', function() {
+        const promise = paginator.next();
+
+        return expect(promise).to.eventually.resolve;
+      });
+
+      it('returns a promise with a response instance', function() {
+        const response = paginator.next();
+
+        return expect(response).to.eventually.be.an.instanceOf(Response);
+      });
+
+      it('increments the current page', function() {
+        expect(paginator.page).to.equal(1);
+
+        const currentPage = paginator.next().then(getCurrentPage);
+
+        return expect(currentPage).to.eventually.equal(2);
```

Comment:
Or you're incrementing after fetching the page, in which case current page makes sense.... Can we get a test that asserts that the response is for the first page?